File: examples/rust/examples/client_websocket_ttl.rs
//! WebSocket TTL operations using ekodb_client library
//!
//! This example demonstrates WebSocket queries with TTL documents using the ekodb_client library.
//!
//! Compare with websocket_ttl.rs to see the difference!

use ekodb_client::{Client, Record};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let ws_url = env::var("WS_BASE_URL").unwrap_or_else(|_| "ws://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created");

    let collection = "client_websocket_ttl_rust";

    // Step 1: Insert test data with TTL via HTTP
    println!("\n=== Insert Test Data with TTL ===");
    let mut record = Record::new();
    record.insert("name", "WebSocket TTL Test");
    record.insert("value", 42);
    let record = record.with_ttl("1h");

    let inserted = client.insert(collection, record).await?;
    println!("✓ Inserted document with TTL: {:?}", inserted.get("id"));

    // Step 2: Query via WebSocket
    println!("\n=== Query via WebSocket ===");
    let ws_client = client.websocket(&ws_url).await?;
    println!("✓ WebSocket connected");

    let records = ws_client.find_all(collection).await?;
    println!("✓ Retrieved {} record(s) via WebSocket", records.len());

    // Show some of the data
    for (i, record) in records.iter().take(3).enumerate() {
        println!("  Record {}: {} fields", i + 1, record.len());
    }

    // Cleanup: Delete the collection
    println!("\n=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection");

    println!("\n✓ WebSocket TTL example completed successfully");
    println!(
        "\n💡 Note: Documents with TTL will automatically expire after the specified duration"
    );

    Ok(())
}

File: examples/kotlin/examples/ClientWebsocketTtl.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout

/**
 * WebSocket TTL example - WebSocket operations with TTL documents
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val wsUrl = dotenv["WS_BASE_URL"] ?: "ws://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_websocket_ttl_example"
    
    println("=== ekoDB Kotlin Client - WebSocket TTL Example ===\n")
    
    try {
        // Step 1: Insert test data with TTL via HTTP
        println("=== Insert Test Data with TTL ===")
        
        val record = Record.new()
            .insert("name", "WebSocket TTL Test")
            .insert("value", 42)
            .insert("created_at", System.currentTimeMillis())
        
        val inserted = client.insertWithTtl(collection, record, "1h")
        println("✓ Inserted document with TTL: ${inserted["id"]}\n")
        
        // Step 2: Query via WebSocket
        println("=== Query via WebSocket ===")
        
        val wsClient = client.websocket(wsUrl)
        wsClient.connect()
        println("✓ WebSocket connected")
        
        try {
            withTimeout(5000) {
                val result = wsClient.findAll(collection)
                println("✓ Retrieved data via WebSocket:")
                println("  $result\n")
            }
        } catch (e: Exception) {
            println("⚠ WebSocket query: ${e.message}\n")
        }
        
        wsClient.close()
        println("✓ WebSocket closed\n")
        
    } finally {
        // Cleanup
        println("=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n✓ WebSocket TTL example completed successfully")
        println("\n💡 Note: Documents with TTL will automatically expire after the specified duration")
    }
}
